Libraries/LibEmail/IMAPClient.cpp

```diff
@@ -0,0 +1,320 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <AK/StringBuilder.h>
+#include <LibEmail/IMAPClient.h>
+#include <ctype.h>
+
+namespace LibEmail {
+
+static ResponseStatus get_response_status(StringView response)
+{
+    if (response.is_empty())
+        return ResponseStatus::Bad;
+
+    const auto right_before_status = response.find_first_of(' ');
+    if (!right_before_status.has_value())
+        return ResponseStatus::Bad;
+
+    if (response.substring_view(right_before_status.value() + 1, 2) == "OK") {
+        return ResponseStatus::Ok;
+    } else if (response.substring_view(right_before_status.value() + 1, 2) == "NO") {
+        return ResponseStatus::No;
+    } else if (response.substring_view(right_before_status.value() + 1, 3) == "BAD") {
+        return ResponseStatus::Bad;
+    } else {
+        return ResponseStatus::FailedToSend;
+    }
+}
+
+static String get_message_field_string(Field value)
+{
+    switch (value) {
+    case Field::InternalDate:
+        return "INTERNALDATE";
+    case Field::Flags:
+        return "FLAGS";
+    case Field::UID:
+        return "UID";
+    case Field::Envelope:
+        return "ENVELOPE";
+    case Field::BodyText:
+        return "BODY[TEXT]";
+    }
+    return "";
+}
+
+IMAPClient::IMAPClient(StringView address, int port)
+    : m_socket{ Core::TCPSocket::construct() }
+{
+    bool connected = m_socket->connect(address, port);
+    dbg() << "Step 1: connection: ";
+    if (connected)
+        dbg() << "Connected successfully";
+    else
+        dbg() << "Failed to connect";
+
+    // Setup thread for receiving/queuing server responses
+    m_receive_thread = LibThread::Thread::construct(
+        [this] {
+            while (true) {
+                const StringBuilder response{ m_socket->receive(1000) };
+                if (response.is_empty()) {
+                    // For some reason servers sometimes send empty messages
+                    continue;
+                }
+                m_queue_lock.lock();
+                m_message_queue.append(response.to_string());
+                m_queue_lock.unlock();
+            }
+            return 0;
+        });
+    m_receive_thread->start();
+}
+
+IMAPClient::~IMAPClient()
+{
+    m_receive_thread->quit();
```

Comment:
How about changing the `while(true)` and doing a clean close? (asking also the `m_socket` to "abort" or something). Killing threads on non-managed languages is not a good idea.